File: regions/system/library/go/resiliency/resiliency_test.go
package resiliency

import (
	"context"
	"errors"
	"fmt"
	"sync/atomic"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestExecute_Success(t *testing.T) {
	policy := ResiliencyPolicy{}
	dec := NewResiliencyDecorator(policy)

	result, err := Execute[int](context.Background(), dec, func() (int, error) {
		return 42, nil
	})

	require.NoError(t, err)
	assert.Equal(t, 42, result)
}

func TestExecute_RetrySuccess(t *testing.T) {
	policy := ResiliencyPolicy{
		Retry: &RetryConfig{
			MaxAttempts: 3,
			BaseDelay:   10 * time.Millisecond,
			MaxDelay:    100 * time.Millisecond,
		},
	}
	dec := NewResiliencyDecorator(policy)

	var counter atomic.Int32
	result, err := Execute[int](context.Background(), dec, func() (int, error) {
		c := counter.Add(1)
		if c < 3 {
			return 0, errors.New("fail")
		}
		return 99, nil
	})

	require.NoError(t, err)
	assert.Equal(t, 99, result)
	assert.Equal(t, int32(3), counter.Load())
}

func TestExecute_MaxRetriesExceeded(t *testing.T) {
	policy := ResiliencyPolicy{
		Retry: &RetryConfig{
			MaxAttempts: 2,
			BaseDelay:   1 * time.Millisecond,
			MaxDelay:    10 * time.Millisecond,
		},
	}
	dec := NewResiliencyDecorator(policy)

	_, err := Execute[int](context.Background(), dec, func() (int, error) {
		return 0, errors.New("always fail")
	})

	require.Error(t, err)
	var rErr *ResiliencyError
	require.True(t, errors.As(err, &rErr))
	assert.Equal(t, "max_retries_exceeded", rErr.Kind)
}

func TestExecute_Timeout(t *testing.T) {
	policy := ResiliencyPolicy{
		Timeout: 50 * time.Millisecond,
	}
	dec := NewResiliencyDecorator(policy)

	_, err := Execute[int](context.Background(), dec, func() (int, error) {
		time.Sleep(1 * time.Second)
		return 42, nil
	})

	require.Error(t, err)
	var rErr *ResiliencyError
	require.True(t, errors.As(err, &rErr))
	assert.Equal(t, "timeout", rErr.Kind)
}

func TestExecute_CircuitBreakerOpens(t *testing.T) {
	policy := ResiliencyPolicy{
		CircuitBreaker: &CircuitBreakerConfig{
			FailureThreshold: 3,
			RecoveryTimeout:  1 * time.Minute,
			HalfOpenMaxCalls: 1,
		},
	}
	dec := NewResiliencyDecorator(policy)

	// Trip the circuit breaker
	for i := 0; i < 3; i++ {
		_, _ = Execute[int](context.Background(), dec, func() (int, error) {
			return 0, errors.New("fail")
		})
	}

	// Next call should fail with circuit open
	_, err := Execute[int](context.Background(), dec, func() (int, error) {
		return 42, nil
	})

	require.Error(t, err)
	var rErr *ResiliencyError
	require.True(t, errors.As(err, &rErr))
	assert.Equal(t, "circuit_open", rErr.Kind)
}

func TestExecute_BulkheadFull(t *testing.T) {
	policy := ResiliencyPolicy{
		Bulkhead: &BulkheadConfig{
			MaxConcurrentCalls: 1,
			MaxWaitDuration:    50 * time.Millisecond,
		},
	}
	dec := NewResiliencyDecorator(policy)

	started := make(chan struct{})
	done := make(chan struct{})

	// Occupy the single bulkhead slot
	go func() {
		_, _ = Execute[int](context.Background(), dec, func() (int, error) {
			close(started)
			<-done
			return 1, nil
		})
	}()

	<-started

	// This should fail with bulkhead full
	_, err := Execute[int](context.Background(), dec, func() (int, error) {
		return 2, nil
	})

	close(done)

	require.Error(t, err)
	var rErr *ResiliencyError
	require.True(t, errors.As(err, &rErr))
	assert.Equal(t, "bulkhead_full", rErr.Kind)
}

func TestCalculateBackoff(t *testing.T) {
	base := 100 * time.Millisecond
	max := 5 * time.Second

	assert.Equal(t, 100*time.Millisecond, calculateBackoff(0, base, max))
	assert.Equal(t, 200*time.Millisecond, calculateBackoff(1, base, max))
	assert.Equal(t, 400*time.Millisecond, calculateBackoff(2, base, max))

	// Should cap at max
	result := calculateBackoff(20, base, max)
	assert.True(t, result <= max, fmt.Sprintf("expected <= %v, got %v", max, result))
}

File: regions/system/library/dart/resiliency/test/resiliency_test.dart
import 'package:test/test.dart';
import 'package:k1s0_resiliency/resiliency.dart';

void main() {
  group('ResiliencyDecorator', () {
    test('正常実行: ポリシーなしで処理が成功すること', () async {
      final decorator = ResiliencyDecorator(const ResiliencyPolicy());
      final result = await decorator.execute(() async => 42);
      expect(result, equals(42));
    });

    test('リトライ: 失敗後に再試行して成功すること', () async {
      final decorator = ResiliencyDecorator(const ResiliencyPolicy(
        retry: RetryConfig(
          maxAttempts: 3,
          baseDelay: Duration(milliseconds: 10),
          maxDelay: Duration(milliseconds: 100),
        ),
      ));

      var counter = 0;
      final result = await decorator.execute(() async {
        counter++;
        if (counter < 3) throw Exception('fail');
        return 99;
      });

      expect(result, equals(99));
      expect(counter, equals(3));
    });

    test('リトライ上限: 最大試行回数を超えると MaxRetriesExceededError が発生すること', () async {
      final decorator = ResiliencyDecorator(const ResiliencyPolicy(
        retry: RetryConfig(
          maxAttempts: 2,
          baseDelay: Duration(milliseconds: 1),
          maxDelay: Duration(milliseconds: 10),
        ),
      ));

      expect(
        () => decorator.execute(() async => throw Exception('always fail')),
        throwsA(isA<MaxRetriesExceededError>()),
      );
    });

    test('タイムアウト: 指定時間内に完了しない場合に TimeoutError が発生すること', () async {
      final decorator = ResiliencyDecorator(const ResiliencyPolicy(
        timeout: Duration(milliseconds: 50),
      ));

      expect(
        () => decorator.execute(() async {
          await Future<void>.delayed(const Duration(seconds: 1));
          return 42;
        }),
        throwsA(isA<TimeoutError>()),
      );
    });

    test('サーキットブレーカー: 連続失敗後にオープン状態へ遷移し CircuitBreakerOpenError が発生すること', () async {
      final decorator = ResiliencyDecorator(const ResiliencyPolicy(
        circuitBreaker: CircuitBreakerConfig(
          failureThreshold: 3,
          recoveryTimeout: Duration(minutes: 1),
          halfOpenMaxCalls: 1,
        ),
      ));

      // しきい値分だけ失敗させてサーキットブレーカーをトリップさせる
      for (var i = 0; i < 3; i++) {
        try {
          await decorator.execute(() async => throw Exception('fail'));
        } catch (_) {}
      }

      // オープン状態では次の呼び出しが即座に拒否される
      expect(
        () => decorator.execute(() async => 42),
        throwsA(isA<CircuitBreakerOpenError>()),
      );
    });

    test('バルクヘッド: 同時実行数が上限に達すると BulkheadFullError が発生すること', () async {
      final decorator = ResiliencyDecorator(const ResiliencyPolicy(
        bulkhead: BulkheadConfig(
          maxConcurrentCalls: 1,
          maxWaitDuration: Duration(milliseconds: 50),
        ),
      ));

      // 唯一のスロットを長時間占有する
      final longRunning = decorator.execute(() async {
        await Future<void>.delayed(const Duration(milliseconds: 500));
        return 1;
      });

      await Future<void>.delayed(const Duration(milliseconds: 10));

      // スロットが埋まっているため次の呼び出しは拒否される
      expect(
        () => decorator.execute(() async => 2),
        throwsA(isA<BulkheadFullError>()),
      );

      await longRunning;
    });
  });

  group('withResiliency', () {
    test('便利関数: ポリシーを適用して正常実行できること', () async {
      final result =
          await withResiliency(const ResiliencyPolicy(), () async => 42);
      expect(result, equals(42));
    });
  });
}
